core/Pos/Core/Genesis/Types.hs

```diff
@@ -155,37 +204,41 @@ bootDustThreshold (GenesisWStakeholders bootHolders) =
     -- even more than 10-15 coins.
     unsafeIntegerToCoin . sum $ map fromIntegral $ toList bootHolders
 
--- | Safe constructor for 'GenesisCoreData'. Throws error if something
+-- | Safe constructor for 'GenesisSpec'. Throws error if something
 -- goes wrong.
-mkGenesisCoreData ::
-       [AddrDistribution]
-    -> Map StakeholderId Word16
+mkGenesisSpec
+    :: GenesisAddrDistr
     -> GenesisDelegation
-    -> Either String GenesisCoreData
-mkGenesisCoreData distribution bootStakeholders delega = do
-    for_ distribution $ \(addrs, distr) -> do
-        -- Every set of addresses should match the stakeholders count
-        unless (fromIntegral (length addrs) == getDistributionSize distr) $
-            Left "mkGenesisCoreData: addressCount != stakeholdersCount \
-                 \for some set of addresses"
-        -- Addresses in each list are distinct (except for CustomBalances)
-        let isCustom = case distr of
-                CustomBalances{} -> True
-                _                -> False
-        unless (isCustom || allDistinct addrs) $
-            Left "mkGenesisCoreData: addresses in some list aren't distinct"
-        unless (all isBootstrapEraDistrAddress addrs) $
-            Left $ "mkGenesisCoreData: there is an address with stake " <>
-                   "distribution different from BootstrapEraDistr"
-
-    -- No address belongs to more than one distribution
-    let addrList = concatMap (ordNub . fst) distribution
-    unless (allDistinct addrList) $
-        Left "mkGenesisCoreData: some address belongs to more than one distr"
+    -> GenesisInitializer
+    -> Either String GenesisSpec
+mkGenesisSpec avvmDistr delega specType = do
+    let addrs = HM.keys $ getAddrDistr avvmDistr
+    unless (all isBootstrapEraDistrAddress addrs) $
+        Left $ "mkGenesisCoreData: there is an address with stake " <>
+                "distribution different from BootstrapEraDistr"
+
     -- All checks passed
-    pure
-        UnsafeGenesisCoreData
-        { gcdAddrDistribution = distribution
-        , gcdBootstrapStakeholders = GenesisWStakeholders bootStakeholders
-        , gcdHeavyDelegation = delega
-        }
+    pure $ UnsafeGenesisSpec avvmDistr delega specType
+
+data ProtocolConstants = ProtocolConstants
+    {
+      -- | Security parameter from paper
+      pcK             :: !Int
+      -- | Magic constant for separating real/testnet
+    , pcProtocolMagic :: !Int32
+      -- | Start time of network (in @Production@ running mode). If set to
+      -- zero, then running time is 2 minutes after build.
+    , pcVssMaxTTL     :: !Word64
+      -- | VSS certificates min timeout to live (number of epochs)
+    , pcVssMinTTL     :: !Word64
+      -- | Threshold of epochs for malicious activity detection
+    }
+data GenesisCanonical = GenesisCanonical
```

Comment:
I'm confused by this name. A Haskell datatype is by nature canonical so the `Canonical` suffix is redundant. It's the JSON representation which may or may not be canonical. Could we rename this to `Genesis`?

Reply:
`GenesisData`?

Reply:
tbh I don't like `GenesisData`, a lot of datatypes can be named `GenesisData` and it doesn't reflect a sense of this datatype.

Reply:
How about `GenesisBlockData`?

In `node/src/Pos//Genesis.hs` there is `data GenesisContext`. Is this `GenesisCanonical` supposed to replace it?